weasyl/commishinfo.py

```diff
@@ -352,100 +357,117 @@ def create_price(userid, price, currency="", settings=""):
-    settings = "".join(i for i in settings if i in "a")
-
-    query = d.execute("SELECT amount_min, amount_max, settings, classid FROM commishprice"
-                      " WHERE (priceid, userid) = (%i, %i)", [price.priceid, userid], options="single")
-
-    if not query:
-        raise WeasylError("priceidInvalid")
-    elif price.amount_min > _MAX_PRICE:
+def edit_price(userid, price, currency, settings, edit_prices):
+    if price.amount_min > _MAX_PRICE:
         raise WeasylError("minamountInvalid")
     elif price.amount_max > _MAX_PRICE:
         raise WeasylError("maxamountInvalid")
     elif price.amount_max and price.amount_max < price.amount_min:
         raise WeasylError("maxamountInvalid")
 
-    argv = []
-    statement = ["UPDATE commishprice SET "]
+    currency = "".join(i for i in currency if i in CURRENCY_CHARMAP)
+    settings = "".join(i for i in settings if i in "a")
+
+    updates = {
+        'settings': currency + settings,
+    }
 
     if price.title:
-        statement.append("%s title = '%%s'" % ("," if argv else ""))
-        argv.append(price.title)
+        updates['title'] = price.title
 
     if edit_prices:
-        if price.amount_min != query[0]:
-            statement.append("%s amount_min = %%i" % ("," if argv else ""))
-            argv.append(price.amount_min)
-
-        if price.amount_max != query[1]:
-            statement.append("%s amount_max = %%i" % ("," if argv else ""))
-            argv.append(price.amount_max)
-
-    statement.append("%s settings = '%%s'" % ("," if argv else ""))
-    argv.append("%s%s" % (currency, settings))
-
-    if not argv:
-        return
-
-    statement.append(" WHERE (priceid, userid) = (%i, %i)")
-    argv.extend([price.priceid, userid])
-
-    d.execute("".join(statement), argv)
+        updates['amount_min'] = price.amount_min
+        updates['amount_max'] = price.amount_max
+
+    cpt = d.meta.tables['commishprice']
+    result = d.engine.execute(
+        cpt.update()
+        .where(cpt.c.priceid == price.priceid)
+        .where(cpt.c.userid == userid)
+        .values(updates)
+    )
+
+    if result.rowcount != 1:
+        raise WeasylError("priceidInvalid")
 
 
 def edit_content(userid, content):
-    if not d.execute("UPDATE commishdesc SET content = '%s' WHERE userid = %i RETURNING userid",
-                     [content, userid], ["element"]):
-        d.execute("INSERT INTO commishdesc VALUES (%i, '%s')", [userid, content])
+    d.engine.execute(
+        "INSERT INTO commishdesc (userid, content)"
+        " VALUES (%(user)s, %(content)s)"
+        " ON CONFLICT (userid) DO UPDATE SET content = %(content)s",
+        user=userid,
+        content=content,
+    )
 
 
 def remove_class(userid, classid):
-    if not d.execute("SELECT EXISTS (SELECT 0 FROM commishclass WHERE (classid, userid) = (%i, %i))",
-                     [d.get_int(classid), userid], ["bool"]):
+    result = d.engine.execute(
```

Comment:
I'm curious, does it breath things if a class and price aren't atomically deleted? I doubt there's much change of a deletion failing since the two deletes are practically right next to each other, but I don't currently know if an isolated commishprice entry would negatively affect anything.

Reply:
I don’t think it would break anything, but it’s undesirable. I’ll put this in a transaction.